engine/resources/docker_container.go

```diff
@@ -0,0 +1,441 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	// ContainerRunning is the running container state.
+	ContainerRunning = "running"
+	// ContainerStopped is the stopped container state.
+	ContainerStopped = "stopped"
+	// ContainerRemoved is the removed container state.
+	ContainerRemoved = "removed"
+
+	// validateCtxTimeout is the length of time, in seconds, before requests
+	// are cancelled in Validate.
+	validateCtxTimeout = 20
+	// checkApplyCtxTimeout is the length of time, in seconds, before requests
+	// are cancelled in CheckApply.
+	checkApplyCtxTimeout = 120
+)
+
+func init() {
+	engine.RegisterResource("docker:container", func() engine.Res { return &DockerContainerRes{} })
+}
+
+// DockerContainerRes is a docker container resource.
+type DockerContainerRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Edgeable
+
+	State      string                     `yaml:"state"`      // running, stopped, or removed
+	Image      string                     `yaml:"image"`      // docker image or image:tag
+	Cmd        []string                   `yaml:"cmd"`        // command(s) to run on the container
+	Env        []string                   `yaml:"env"`        // environment variables, e.g. ["VAR"="val",],
+	Ports      map[string]map[int64]int64 `yaml:"ports"`      // port bindings, e.g. {"tcp" => {80 => 8080},},
+	APIVersion string                     `yaml:"apiversion"` // optional api version, e.g. "v1.39"
+
+	Force bool `yaml:"force"` // force destroy and redeploy if image is incorrect
+
+	client *client.Client // docker api client
+
+	init *engine.Init
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *DockerContainerRes) Default() engine.Res {
+	return &DockerContainerRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *DockerContainerRes) Validate() error {
+	ctx, cancel := context.WithTimeout(context.Background(), validateCtxTimeout*time.Second)
+	defer cancel()
+
+	// validate state
+	if obj.State != ContainerRunning && obj.State != ContainerStopped && obj.State != ContainerRemoved {
+		return fmt.Errorf("state must be running, stopped or removed")
+	}
+
+	// TODO: should this stuff happen in validate or do we make sure there's no
+	// whitespace, and let it fail in CheckApply?
+	//
+	// set up temporary docker client to validate image
+	c, err := client.NewClient(client.DefaultDockerHost, obj.APIVersion, nil, nil)
```

Comment:
Let's not make any connections in Validate()
Keep it "cheap" so we can casually use Validate() in many places.
Add these checks to Init() or CheckApply. Init seems more likely.
Of course, anything you can do without the connection should be done in Validate.
Why? Well we might consider adding Validate() calls from the lang in order to statically check things that are determined statically as a compiler optimization. Eg: If you asked for state => "wrong", since those are all constants we can catch this at compile time... Only if Validate is cheap =D